TLM/TLM/Util/Extensions/NetSegmentExtensions.cs

```diff
@@ -374,7 +381,10 @@ public static class NetSegmentExtensions {
 
                 laneList.Sort(CompareLanePositionsFun);
             }
+
             return laneList;
         }
+
+        private static readonly List<LanePos> EmptyLaneList = new(0);
```

Comment:
`ReadOnlyCollection<T>`?

Reply:
The only reason I marked the `EmptyLaneList` a readonly member is just to convey intent that it should not be altered.

To use `ReadOnlyCollection` I would need to make `GetSortedLanes()` return a `ReadOnlyCollection`.

I don't see any benefit of doing that:

- `GetSortedLanes()` is internal to TM:PE
- We never modify the returned list or its members
- `ReadOnlyCollection` has a small performance hit
- We use `GetSortedLanes()` in overlay hotpaths

Reply:
> `ReadOnlyCollection` has a small performance hit

What performance hit? It's collection that throws exception when you try to modify it, class itself is based on `List<>`.

> We never modify the returned list or its members

I think you wanted to say _"In our **current code** we don't modify it"_, but nothing is blocking us and worse, it will break the empty list you return causing quite hard bug to spot. 
Setting a field which holds the class instance as `readonly` does not block the value (instance) from modification, it only prevents from changing the instance stored in that field. It's very useful if you want to ensure that instance set to the field will never change (but not its contents). That's why suggested real read-only list. I'm not sure why the method even returns `IList` - it could be anything, so why not `ReadOnlyCollection<>`? It does not violate return type in any way 🙃 

> We use `GetSortedLanes()` in overlay hotpaths

So if we care that much about performance then the method should be rewritten because it generates garbage for non-zero element results. First it creates a list with max possible size then trims if internal capacity is bigger than the actual number of elements added.

So summarizing previous impl. was far from perfect, now it's better but still with the room for improvements. I think we've talked with @Elesbaan70 on discord about possibilities for caching sorted lanes or segment lanes in general so we may return to this thing in the future.

Reply:
> What performance hit?

It's a wrapper for a list (just as list is wrapper for strongly typed array) so incurs some overheads.

https://www.infoq.com/articles/For-Each-Performance/

![image](https://user-images.githubusercontent.com/1386719/162471662-b52fb2cb-164a-4e1b-8005-b18ca0f755b5.png)

That being said, I assume this pales in to insignificance compared to the time taken to iterate and process the lanes, etc., so probably not worth worrying about.

> I'm not sure why the method even returns IList

Same here. Although it's been used for the `Record` stuff in asset editor, etc., so there could be implications in changing it? (Assuming we can convert a `ReadOnlyCollection` to a `List` to satisfy the return type of several methods, it wouldn't be a problem, but I've not tested that assumption - would we need to create a new list based on the `ReadOnlyCollection` and would that cause any issues?)

> possibilities for caching sorted lanes or segment lanes in general

I doubt we could do trivial caching as there are lots of parameter combos - eg. direction, lane type, vehicle type, etc. We would probably need to extend `LanePos` struct to include some additional fields (lane type, vehicle type, etc) and cache unsorted lists of those, and make some filterable/sortable enumerators?

Also, would we do it for all segements or just those in view? In many cases `GetSortedLanes()` is being used in the code that renders overlays so that would merit some caching. I've no idea if TTL uses `GetSortedlanes()` during gameplay but, if so, that would be even more important to cache.

Reply:
I don't trust that benchmark at all. One additional method call to get value by index (indexer is calling indexer of internal list) shouldn't slow down iteration by ~6x, come on😄 

Maybe it was not 100% clear but I meant the `readonly` field to be `ReadOnlyCollection` for now (only) so it will ensure that our 0 element cache list will always be empty and prevent weird bugs.

As of caching, we could cache containers (arrays or lists - pull and return from pool) instead of spamming with new collections each time so in that case having read-only thing could help a little bit. Anyways, it's a thing to think about in the future, out of scope of this PR 😉

Reply:
`GetSortedLanes()` can't return a `ReadOnlyCollection` unless I alter its return type to same (as far as I can tell) in which case non-empty lane lists will also need to be `ReadOnlyCollection` (maybe I'm missing something?)

I can easily convert the `GetSortedLanes()` to return `ReadOnlyCollection` but ideally it requires #1521 merging first as that removes some explicit `IList` variables at the call sites and I might need to convert some of the returned `ReadOnlyCollection` to `List` in the `Record` classes.

Reply:
Wait.. method already returns `IList<>`, don't change that (I assume you won't to change that? 🤔). It will accept `ReadOnlyCollection<>` as return object. Nothing else is required change. It's just simple polymorphism.

`private static readonly ReadOnlyCollection<LanePos> EmptyLaneList = new ReadOnlyCollection<LanePos>(new List<LanePos>(0));`

If you plan changing IList to something else (maybe I missed that part) ignore everything what I wrote about changes in this PR.

Reply:
for some reason that didn't work when I tried previously :/ anyway, it's working now so I've sent commit to make the emty list use readonly collection

Reply:
> I think we've talked with @Elesbaan70 on discord about possibilities for caching sorted lanes or segment lanes in general so we may return to this thing in the future.

The displaced lane detection caches its results. The same cache could also keep a sorted collection of `NetInfo.Lane`.

Some time soon I'll be creating a pull request that caches lane index information for `NetLane` so we don't have to find it in a loop every time we need it. It does this using an `ExtLaneManager`, which could also be used in conjunction with `ExtSegmentManager` to cache a sorted linked list of `NetLane` if that's useful.

Reply:
@Elesbaan70 sounds good! Looking forward to it!

Reply:
It's delayed because I don't have a good way to test it thoroughly until I merge @kianzarrin's rewritten `LaneConnectionManager` with my displaced lane code. 😅